### PR TITLE
use some bug_features in backout model

### PR DESCRIPTION
**Pull request by chidauri**
For #577

### CODE REVIEW

**Comment by marco-c** [comment]
Can you post the (full) results before/after your change?

**Comment by chidauri** [comment]
____________________________________________________________After changes______________________________________________________________________________

35473 commits were backed out
363697 commits were not backed out
X: (399170, 1026353), y: (399170,)
Cross Validation scores:
Accuracy: f0.6754237164967368 (+/- 0.0068275143218936496)
Precision: f0.17796056601799792 (+/- 0.002994576181311574)
Recall: f0.7324106652925543 (+/- 0.0023711790743826022)
X_train: (63874, 1026353), y_train: (63874,)
X_test: (39917, 1026353), y_test: (39917,)
No confidence threshold - 39917 classified
[[ 2578   958]
 [11880 24501]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.18      0.73      0.67      0.29      0.70      0.49      3536
          0       0.96      0.67      0.73      0.79      0.70      0.49     36381

avg / total       0.89      0.68      0.72      0.75      0.70      0.49     39917


Confidence threshold > 0.6 - 27604 classified
[[ 1960   520]
 [ 6692 18432]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.23      0.79      0.73      0.35      0.76      0.58      2480
          0       0.97      0.73      0.79      0.84      0.76      0.58     25124

avg / total       0.91      0.74      0.79      0.79      0.76      0.58     27604


Confidence threshold > 0.7 - 16086 classified
[[ 1170   205]
 [ 2777 11934]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.30      0.85      0.81      0.44      0.83      0.69      1375
          0       0.98      0.81      0.85      0.89      0.83      0.69     14711

avg / total       0.92      0.81      0.85      0.85      0.83      0.69     16086


Confidence threshold > 0.8 - 7690 classified
[[ 420   54]
 [ 686 6530]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.38      0.89      0.90      0.53      0.90      0.80       474
          0       0.99      0.90      0.89      0.95      0.90      0.80      7216

avg / total       0.95      0.90      0.89      0.92      0.90      0.80      7690


Confidence threshold > 0.9 - 3794 classified
[[  47    2]
 [  36 3709]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.57      0.96      0.99      0.71      0.97      0.95        49
          0       1.00      0.99      0.96      0.99      0.97      0.95      3745

avg / total       0.99      0.99      0.96      0.99      0.97      0.95      3794


________________________________________________________without changes_____________________________________________________________


35473 commits were backed out
363697 commits were not backed out
X: (399170, 1019320), y: (399170,)
Cross Validation scores:
Accuracy: f0.6681252406342949 (+/- 0.00798929570758301)
Precision: f0.17463934444926482 (+/- 0.003846040864042139)
Recall: f0.7334125536186675 (+/- 0.008836331972903415)
X_train: (63874, 1019320), y_train: (63874,)
X_test: (39917, 1019320), y_test: (39917,)
No confidence threshold - 39917 classified
[[ 2577   959]
 [12323 24058]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.17      0.73      0.66      0.28      0.69      0.49      3536
          0       0.96      0.66      0.73      0.78      0.69      0.48     36381

avg / total       0.89      0.67      0.72      0.74      0.69      0.48     39917


Confidence threshold > 0.6 - 27330 classified
[[ 1923   511]
 [ 6943 17953]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.22      0.79      0.72      0.34      0.75      0.57      2434
          0       0.97      0.72      0.79      0.83      0.75      0.57     24896

avg / total       0.91      0.73      0.78      0.78      0.75      0.57     27330


Confidence threshold > 0.7 - 15507 classified
[[ 1103   202]
 [ 2774 11428]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.28      0.85      0.80      0.43      0.82      0.68      1305
          0       0.98      0.80      0.85      0.88      0.82      0.68     14202

avg / total       0.92      0.81      0.84      0.85      0.82      0.68     15507


Confidence threshold > 0.8 - 7253 classified
[[ 362   53]
 [ 587 6251]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.38      0.87      0.91      0.53      0.89      0.79       415
          0       0.99      0.91      0.87      0.95      0.89      0.80      6838

avg / total       0.96      0.91      0.87      0.93      0.89      0.80      7253


Confidence threshold > 0.9 - 3778 classified
[[  28    1]
 [  19 3730]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.60      0.97      0.99      0.74      0.98      0.96        29
          0       1.00      0.99      0.97      1.00      0.98      0.96      3749

avg / total       1.00      0.99      0.97      1.00      0.98      0.96      3778

@marco-c

**Comment by chidauri** [comment]
Trained on 132000 bugs,which were associated to the commits only, above that my system crashed :(

File: bugbug/models/backout.py [hunk 1]
```diff
@@ -9,13 +9,13 @@
 from sklearn.feature_extraction import DictVectorizer
 from sklearn.pipeline import Pipeline
 
-from bugbug import commit_features, feature_cleanup, repository
+from bugbug import bug_features, commit_features, feature_cleanup, repository
 from bugbug.model import CommitModel
 
 
 class BackoutModel(CommitModel):
     def __init__(self, lemmatization=False):
```
Comment (marco-c) [inline bugbug/models/backout.py]: I'd add an option here (`bug_data`), so the user can decide whether to use bug features or not.

File: bugbug/model.py [hunk 2]
```diff
@@ -286,6 +286,8 @@ def items_gen(self, classes):
                 if commit["node"] in classes
             )
 
+            bugzilla.download_bugs(all_bug_ids)
```
Comment (marco-c) [inline bugbug/model.py]: We need to move this somewhere else, model.py is not the right place.

We'll probably have to move it in the bug_retriever script.

Comment (chidauri) [reply bugbug/model.py]: @marco-c if i add this to (get_all_bug_ids) in labels.py , we don't need to call it, bug_retriever script will automatically take care of it, right?

Comment (marco-c) [reply bugbug/model.py]: Yes

File: bugbug/labels.py [hunk 3]
```diff
@@ -42,4 +42,6 @@ def get_all_bug_ids():
 
             bug_ids.update([int(row["bug_id"]) for row in reader])
 
+    bug_ids.update(commit["bug_id"] for commit in repository.get_commits())
+
     return list(bug_ids)
```
Comment (marco-c) [inline bugbug/labels.py]: What is the length of this list before/after your changes? I want to make sure we're not adding too many bugs here

File: bugbug/models/backout.py (hunk 1 above)
Comment (marco-c) [reply bugbug/models/backout.py]: You should also actually only use the bug features when "bug_data" is True.

Comment (marco-c) [reply bugbug/models/backout.py]:
```Python
if bug_data:
    feature_extractors += [
            bug_features.product(),
            bug_features.component(),
            ...
    ]
```

File: bugbug/labels.py (hunk 3 above)
Comment (chidauri) [reply bugbug/labels.py]: you're right, previously 12967 bugs and after changes 242435 bugs, so what should we do?
PS. The downloading is very slow and takes like 3 hours approx. :(

Comment (marco-c) [reply bugbug/labels.py]: A possible option is to restrict commits to the past 2 years (or 2 years and a half).
Something like:
```Python
start_date = datetime.now() - relativedelta(years=2, months=6)
bug_ids.update(commit["bug_id"] for commit in repository.get_commits() if commit["bug_id"] and dateutil.parser.parse(commit['pushdate']) <= start_date)
```

What is the number of bugs then?

File: bugbug/models/backout.py (hunk 1 above)
Comment (chidauri) [reply bugbug/models/backout.py]: Aren't these lines already taking care of that?
https://github.com/mozilla/bugbug/blob/b456bd3b2d00aad453b42cc8761fc04fbf0e3866/bugbug/commit_features.py#L178-L181

File: bugbug/labels.py (hunk 3 above)
Comment (chidauri) [reply bugbug/labels.py]: 179964

File: bugbug/models/backout.py (hunk 1 above)
Comment (marco-c) [reply bugbug/models/backout.py]: Yes, but that check is not specifically for this, so we can't rely on it without making the code harder to maintain.
If we rely on checks that are there for other reasons, we introduce hidden dependencies in the code, which make it harder to modify later (what if somebody removes the "if not commit[..." check because they don't see that it is useful for the "backout" model?).

File: bugbug/labels.py (hunk 3 above)
Comment (marco-c) [reply bugbug/labels.py]: Can you check how many of these bugs are not already there in the bugs DB?

Comment (marco-c) [reply bugbug/labels.py]: If the two sets are disjoint, we can't really add 180000 bugs, but if the two sets share a lot of bugs maybe the 180000 will become way less.

Comment (chidauri) [reply bugbug/labels.py]: 166000 are not there in bugs db.(only 13000 are)

Comment (marco-c) [reply bugbug/labels.py]: Oh wow, I was not expecting such a big number!

Comment (marco-c) [reply bugbug/labels.py]: Did you use my snippet as-is? There was an error (`<=` instead of `>=`). If you fix that, the number is 23k.

Comment (chidauri) [reply bugbug/labels.py]: I had a feeling that it should be >= but i used the snippet as is :(

File: bugbug/models/backout.py [hunk 4]
```diff
@@ -9,13 +9,13 @@
...
-from bugbug import commit_features, feature_cleanup, repository
+from bugbug import bug_features, commit_features, feature_cleanup, repository
 from bugbug.model import CommitModel
 
 
 class BackoutModel(CommitModel):
-    def __init__(self, lemmatization=False):
-        CommitModel.__init__(self, lemmatization)
+    def __init__(self, lemmatization=False, bug_data=False):
+        CommitModel.__init__(self, lemmatization, bug_data)
```
Comment (marco-c) [inline bugbug/models/backout.py]: When bug_data is True, we should only get the commits for which we have bug_data, so in get_labels we should have the same date check as in bug_retriever

Comment (chidauri) [reply bugbug/models/backout.py]: Done.

Comment (marco-c) [reply bugbug/models/backout.py]: I think there are no dependencies missing, you are not using any new package.

**Comment by marco-c** [comment]
Looks good, but tests are failing because the commits in the commits.json mock file (tests/fixtures/commits.json) don't contain the "pushdate" value. I think the test failure will be fixed by adding it to all commits.

You can run the tests locally by running:
```Bash
python3 -m pytest tests/test_*
```

**Comment by chidauri** [comment]
I added a random pushdate to all commits

File: tests/fixtures/commits.json [hunk 5]
```diff
@@ -1,5 +1,5 @@
...
+{"author": "Mike Hommey <mh+mozilla@glandium.org>", "desc": "Bug 1488307 - Build the LLVM gold plugin. r=mshal\n\nAs much as it's tempting to use LLD for LTO, it still causes some\nsubtle problems with the build, and it's still better to keep using\nBFD ld for the time being. Doing so requires the gold plugin, which\nonly requires to pass cmake the directory where the binutils headers\nare, and they are part of the gcc toolchain headers.\n\nDifferential Revision: https://phabricator.services.mozilla.com/D4896", "date": "2018-08-30 01:55:49", "pushdate": "2018-09-05 09:45:59", "bug_id": 1488307, "ever_backedout": true, "added": 2, "test_added": 0, "deleted": 0, "test_deleted": 0, "files_modified_num": 1, "types": ["Python"], "components": ["Firefox Build System::General"], "author_experience": 1147, "author_experience_90_days": 125, "author_email": "mh+mozilla@glandium.org", "node": "e2a02b08089b0bd0c18ceac0b2eb1e3888d56dc2"}
```
Comment (chidauri) [inline tests/fixtures/commits.json]: @marco-c these last two commits have their "ever_backedout" values wrongly swapped, can that be a problem?

File: bugbug/models/backout.py [hunk 6]
```diff
@@ -71,7 +89,17 @@ def __init__(self, lemmatization=False):
...
 
+        six_months_ago = datetime.utcnow() - relativedelta(months=6)
+        two_years_and_six_months_ago = six_months_ago - relativedelta(years=2)
+
         for commit_data in repository.get_commits():
+            if self.bug_data:
+                pushdate = dateutil.parser.parse(commit_data["pushdate"])
+                if not (
+                    pushdate >= two_years_and_six_months_ago
+                    and pushdate <= six_months_ago
```
Comment (marco-c) [inline bugbug/models/backout.py]: We don't need this side here, just pushdate >= two_years_and_six_months_ago.
Also add a comment saying that we restrict the pushdata when we use the bug data because we don't have bugs for all commits and that we need to keep the dates here and in bug_retriever in sync.

File: tests/fixtures/commits.json (hunk 5 above)
Comment (marco-c) [reply tests/fixtures/commits.json]: What do you mean by "wrongly swapped"?

**Comment by marco-c** [comment]
After downloading the bugs, can you check if the accuracy changed?

File: tests/fixtures/commits.json (hunk 5 above)
Comment (chidauri) [reply tests/fixtures/commits.json]: @marco-c means the one which has "ever_backedout" =True is actually False in data/commits.json , and vice-versa

Comment (marco-c) [reply tests/fixtures/commits.json]: Ah OK, it shouldn't be a problem, but better to fix it.

**Comment by chidauri** [comment]
> After downloading the bugs, can you check if the accuracy changed?

___________________________At current settings with downloaded bugs____________________
15661 commits were backed out
111292 commits were not backed out
X: (126953, 564359), y: (126953,)
Cross Validation scores:
Accuracy: f0.6924302300699183 (+/- 0.004268721905037634)
Precision: f0.24114152139062267 (+/- 0.0024677271627321414)
Recall: f0.7008048304043882 (+/- 0.01568485997561132)
X_train: (28062, 564359), y_train: (28062,)
X_test: (12696, 564359), y_test: (12696,)
No confidence threshold - 12696 classified
[[1163  467]
 [3521 7545]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.25      0.71      0.68      0.37      0.70      0.49      1630
          0       0.94      0.68      0.71      0.79      0.70      0.48     11066

avg / total       0.85      0.69      0.71      0.74      0.70      0.49     12696


Confidence threshold > 0.6 - 8160 classified
[[ 836  225]
 [1778 5321]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.32      0.79      0.75      0.45      0.77      0.59      1061
          0       0.96      0.75      0.79      0.84      0.77      0.59      7099

avg / total       0.88      0.75      0.78      0.79      0.77      0.59      8160


Confidence threshold > 0.7 - 4135 classified
[[ 427   73]
 [ 678 2957]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.39      0.85      0.81      0.53      0.83      0.70       500
          0       0.98      0.81      0.85      0.89      0.83      0.69      3635

avg / total       0.90      0.82      0.85      0.84      0.83      0.69      4135


Confidence threshold > 0.8 - 1361 classified
[[ 139   11]
 [ 133 1078]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.51      0.93      0.89      0.66      0.91      0.83       150
          0       0.99      0.89      0.93      0.94      0.91      0.82      1211

avg / total       0.94      0.89      0.92      0.91      0.91      0.82      1361


Confidence threshold > 0.9 - 507 classified
[[  4   0]
 [  1 502]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.80      1.00      1.00      0.89      1.00      1.00         4
          0       1.00      1.00      1.00      1.00      1.00      1.00       503

avg / total       1.00      1.00      1.00      1.00      1.00      1.00       507

**Comment by marco-c** [comment]
Nice, sounds like the precision increased by ~7%.

**Comment by chidauri** [comment]
> Nice, sounds like the precision increased by ~7%.

yeah, significant improvement!

File: bugbug/models/backout.py [hunk 7]
```diff
@@ -71,7 +89,16 @@ def __init__(self, lemmatization=False):
...
 
+        two_years_and_six_months_ago = datetime.utcnow() - relativedelta(
+            years=2, months=6
+        )
+
+        # when using bug_data we restrict the pushdate to get only those commits for which we have bug_data by keeping the dates in sync with bug_retriever
         for commit_data in repository.get_commits():
+            if self.bug_data:
+                pushdate = dateutil.parser.parse(commit_data["pushdate"])
+                if not pushdate >= two_years_and_six_months_ago:
```
Comment (marco-c) [inline bugbug/models/backout.py]: This is a bit backwards, better to use < instead of not >= :P
